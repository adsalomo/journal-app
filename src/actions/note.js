import Swal from "sweetalert2";
import { db } from "../firebase/firebase-config";
import { fileUpload } from "../helpers/fileUpload";
import { loadNotes } from "../helpers/loadNotes";
import { types } from '../types/types';

export const startNewNote = () => {
    return async (dispatch, getState) => {

        const uid = getState().auth.uid;

        const newNote = {
            title: '',
            body: '',
            date: new Date().getTime(),
        };

        const doc = await db.collection(`${uid}/journal/note`).add(newNote);

        dispatch(activeNote(doc.id, newNote));
        dispatch(addNewNote(doc.id, newNote));
    };
};

export const activeNote = (id, note) => {
    return {
        type: types.noteActive,
        payload: {
            id,
            ...note,
        }
    };
};

export const addNewNote = (id, note) => {
    return {
        type: types.noteAddNew,
        payload: {
            id, ...note,
        },
    }
};

export const startLoadingNote = (uid) => {
    return async (dispatch) => {
        const notes = await loadNotes(uid);
        dispatch(setNotes(notes));
    };
}

export const setNotes = (notes) => {
    return {
        type: types.noteLoad,
        payload: notes,
    };
};

export const startSaveNote = (note) => {
    return async (dispatch, getState) => {

        const { uid } = getState().auth;

        // FIREBASE NO SOPORTA UNDEFINED, LA URL ESTA HASTA LA CLASE 263 SIN LA URL
        if (!note.url) {
            delete note.url;
        }

        const noteToFireStore = { ...note };
        delete noteToFireStore.id;

        await db.doc(`${uid}/journal/note/${note.id}`).update(noteToFireStore);

        dispatch(refreshNote(note.id, noteToFireStore));

        Swal.fire('Saved', note.title, 'success');
    }
};

export const refreshNote = (id, note) => {
    return {
        type: types.noteUpdated,
        payload: {
            id,
            note: {
                id,
                ...note
            }
        }
    }
};

export const startUploading = (file) => {
    return async (dispatch, getState) => {
        const { active: activeNote } = getState().note;

        Swal.fire({
            title: 'Uploading...',
            text: 'Please wait...',
            allowOutsideClick: false,
        });

        Swal.showLoading();

        const fileUrl = await fileUpload(file);
        activeNote.url = fileUrl;

        dispatch(startSaveNote(activeNote));

        Swal.close();
    }
};

export const startDeleting = (id) => {
    return async (dispatch, getState) => {

        const uid = getState().auth.uid;
        await db.doc(`${uid}/journal/note/${id}`).delete();

        dispatch(deleteNote(id));
    }
};

export const deleteNote = (id) => {
    return {
        type: types.noteDeleteNote,
        payload: id,
    }
};

export const noteLogout = () => {
    return {
        type: types.noteLogoutCleaning,
    }
};

