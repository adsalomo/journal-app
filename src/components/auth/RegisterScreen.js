import React from 'react';
import { Link } from 'react-router-dom';

export const RegisterScreen = () => {
    return (
        <>
            <h3 className="auth__title mb-2">Register</h3>

            <form>
                <input
                    type="text"
                    placeholder="Name"
                    name="name"
                    className="auth__input"
                    autoComplete="off"
                />

                <input
                    type="text"
                    placeholder="Email"
                    name="email"
                    className="auth__input"
                    autoComplete="off"
                />

                <input
                    type="passwrod"
                    placeholder="Password"
                    name="password"
                    className="auth__input"
                />

                <input
                    type="passwrod"
                    placeholder="Confirm password"
                    name="password2"
                    className="auth__input"
                />

                <button
                    className="btn btn-primary btn-block mb-2"
                    type="submit">
                    Register
                </button>

                <Link to="/auth/login" className="link">
                    Already Registered?
                </Link>
            </form>
        </>
    )
}
