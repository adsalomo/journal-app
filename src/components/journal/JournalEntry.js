import React from 'react'

export const JournalEntry = () => {
    return (
        <div className="journal__entry pointer">

            <div
                className="journal__entry-picture"
                style={{
                    backgroundSize: 'cover',
                    backgroundImage: 'url(https://computacioninteractiva.com/wp-content/uploads/2019/07/INVERTIR-EN-TECNOLOGIA.png)'
                }}
            ></div>

            <div className="journal__entry-body">
                <p className="journal__entry-title">Un nuevo dia</p>
                <p className="journal__entry-content">Un nuevo dia</p>
            </div>

            <div className="journal__entry-date-box">
                <span>Monday</span>
                <h4>28</h4>
            </div>
        </div>
    )
}
