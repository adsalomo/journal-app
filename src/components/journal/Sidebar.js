import React from 'react'
import { JournalEntries } from './JournalEntries'

export const Sidebar = () => {
    return (
        <aside className="journal__sidebar">

            <div className="journal__sidebar-navbar">
                <h3 className="mt-2">
                    <li className="far fa-moon"></li>
                    <span>Elena</span>
                </h3>

                <button className="btn">
                    Logout
                </button>
            </div>

            <div className="journal__new-entry">
                <li className="far fa-calendar-plus fa-5x"></li>
                <p className="mt-2">New entry</p>
            </div>

            <JournalEntries />

        </aside>
    )
}
