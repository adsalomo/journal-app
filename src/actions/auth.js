import { types } from '../types/types';
import { firebase, googleAuthProvide } from '../firebase/firebase-config';
import { uiFinishLoading, uiStartLoading } from './ui';
import Swal from 'sweetalert2'
import { noteLogout } from './note';

// middleware
export const startLoginEmailPassword = (email, password) => {
    return (dispatch) => {

        dispatch(uiStartLoading());

        firebase.auth().signInWithEmailAndPassword(email, password)
            .then(({ user }) => {
                dispatch(uiFinishLoading());

                dispatch(login(user.uid, user.displayName));
            }, e => {
                dispatch(uiFinishLoading());
                console.log(e);
                Swal.fire('Error', e.message, 'error');
            });

    }
};

// middleware
export const startRegister = (name, email, password) => {
    return (dispatch) => {

        firebase.auth().createUserWithEmailAndPassword(email, password)
            .then(async ({ user }) => {
                await user.updateProfile({ displayName: name });
                dispatch(login(user.uid, user.displayName));
            }, e => {
                console.log(e);
                Swal.fire('Error', e.message, 'error');
            });

    }
};

// middleware
export const startGoogleLogin = () => {
    return (dispatch) => {

        firebase.auth().signInWithPopup(googleAuthProvide)
            .then(({ user }) => {
                dispatch(login(user.uid, user.displayName));
            }, e => {
                console.log(e);
            });

    }
};

export const login = (uid, displayName) => {
    return {
        type: types.login,
        payload: {
            uid,
            displayName,
        }
    }
};

// middleware
export const startLogout = () => {
    return async (dispatch) => {

        await firebase.auth().signOut();
        dispatch(logout());
        dispatch(noteLogout());
    }
}

export const logout = () => {
    return {
        type: types.logout,
    }
}

