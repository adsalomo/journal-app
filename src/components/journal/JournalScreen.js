import React from 'react';
import { NoteScreen } from '../notes/NoteScreen';
import { Sidebar } from './Sidebar';
import { useSelector } from 'react-redux';
import { NothingSelected } from './NothingSelected';

export const JournalScreen = () => {
    const { active } = useSelector(state => state.note);

    return (
        <div
            className="journal__main-content animate__animated animate__fadeIn animate__faster"
        >

            <Sidebar />

            <main>

                {
                    active ? <NoteScreen /> : <NothingSelected />
                }

            </main>
        </div>
    )
}
