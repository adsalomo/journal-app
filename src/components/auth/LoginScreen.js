import React from 'react';
import { Link } from 'react-router-dom';
import { useForm } from '../../hooks/useForm';
import { useDispatch, useSelector } from 'react-redux';
import { startGoogleLogin, startLoginEmailPassword } from '../../actions/auth';
import validator from 'validator';
import { removeError, setError } from '../../actions/ui';

export const LoginScreen = () => {

    const dispatch = useDispatch();
    const { msgError, loading } = useSelector(state => state.ui);

    const [formValues, handleInputChange] = useForm({
        email: '',
        password: '',
    });

    const { email, password } = formValues;

    const handleLogin = (e) => {
        e.preventDefault();

        if (!isFormValid()) {
            return;
        }

        dispatch(startLoginEmailPassword(email, password));
    }

    const handleGoogleLogin = () => {
        dispatch(startGoogleLogin());
    }

    const isFormValid = () => {
        if (!validator.isEmail(email)) {
            dispatch(setError('Email not valid'));
            return false;
        } else if (password.length < 5) {
            dispatch(setError('Password not valid'));
            return false;
        }

        dispatch(removeError());

        return true;
    }

    return (
        <>
            <h3 className="auth__title mb-2">Login</h3>

            {
                msgError &&
                <div className="auth__alert-error">
                    {msgError}
                </div>
            }

            <form
                onSubmit={handleLogin}
                className="animate__animated animate__fadeIn animate__faster"
            >
                <input
                    type="text"
                    placeholder="email"
                    name="email"
                    className="auth__input"
                    autoComplete="off"
                    value={email}
                    onChange={handleInputChange}
                />

                <input
                    type="password"
                    placeholder="password"
                    name="password"
                    className="auth__input"
                    value={password}
                    onChange={handleInputChange}
                />

                <button
                    className="btn btn-primary btn-block"
                    disabled={loading}
                    type="submit">
                    Login
                </button>

                <div className="auth__social-network">
                    <p>Login social network</p>
                    <div
                        className="google-btn"
                        onClick={handleGoogleLogin}
                    >
                        <div className="google-icon-wrapper">
                            <img className="google-icon" src="https://upload.wikimedia.org/wikipedia/commons/5/53/Google_%22G%22_Logo.svg" alt="google button" />
                        </div>
                        <p className="btn-text">
                            <b>Sign in with google</b>
                        </p>
                    </div>
                </div>

                <Link to="/auth/register" className="link">
                    Create new account
                </Link>
            </form>
        </>
    )
}
