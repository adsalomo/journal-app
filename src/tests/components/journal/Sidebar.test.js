import { mount } from "enzyme";
import { Provider } from "react-redux";
import configureStore from 'redux-mock-store';
import thunk from 'redux-thunk';
import { startLogout } from "../../../actions/auth";
import { startNewNote } from "../../../actions/note";
import { Sidebar } from "../../../components/journal/Sidebar";

jest.mock('../../../actions/auth', () => {
    return {
        startLogout: jest.fn(),
    }
});

jest.mock('../../../actions/note', () => {
    return {
        startNewNote: jest.fn(),
    }
});

const middlewares = [thunk];
const mockStore = configureStore(middlewares);


const initialState = {
    auth: {
        uid: 'ABC',
        name: 'Elena'
    },
    note: {
        notes: []
    }
};

let store = mockStore(initialState);
store.dispatch = jest.fn();

describe('Pruebas <Sidebar />', () => {

    const wrapper = mount(
        <Provider store={store}>
            <Sidebar />
        </Provider>
    );

    test('Debe de mostrarse correctamente', () => {
        expect(wrapper).toMatchSnapshot();
    });

    test('Debe de llamar la accion startLogout', () => {
        wrapper.find('button').prop('onClick')();
        expect(startLogout).toHaveBeenCalled();
    });

    test('Debe de llamar la accion startNewNote', () => {
        wrapper.find('.journal__new-entry').prop('onClick')();
        expect(startNewNote).toHaveBeenCalled();
    });


});
