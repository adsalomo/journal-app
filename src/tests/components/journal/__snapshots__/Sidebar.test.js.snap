// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Pruebas <Sidebar /> Debe de mostrarse correctamente 1`] = `
<aside
  className="journal__sidebar"
>
  <div
    className="journal__sidebar-navbar"
  >
    <h3
      className="mt-2"
    >
      <li
        className="far fa-moon"
      />
      <span>
        Elena
      </span>
    </h3>
    <button
      className="btn"
      onClick={[Function]}
    >
      Logout
    </button>
  </div>
  <div
    className="journal__new-entry"
    onClick={[Function]}
  >
    <li
      className="far fa-calendar-plus fa-5x"
    />
    <p
      className="mt-2"
    >
      New entry
    </p>
  </div>
  <div
    className="journal__entries"
  />
</aside>
`;
